### ISSUE TITLE
pycdf attribute writing speed

### ISSUE THREAD

**Issue by jtniehof** [spacepy/spacepy#89]
Writing an entry to an attribute can be awfully slow. Direct assignment is slow from type guessing, but even calling .new() is not fast. I think part of the issue is checking to see if it exists several times..."new" explicitly checks for existence so it can raise a KeyError, and then it calls _get_or_create which tries again to see if it's in the file. I think _get_or_create can be replaced with a direct call to the AttrType constructor to avoid at least one of these. There's also lots of calls to global_scope that might be overkill (and honestly the SCOPE attribute should ensure that always works; those should maybe be class attributes anyhow).

This isn't an issue when just writing one attribute, but we've got some files with hundreds of variables and dozens of attributes each, and all these checks require passing through the ctypes barrier into the CDF library.

I probably need to implement a reasonable benchmark that's thinner than the examples I've been running into and work against that.

**Comment by jtniehof** [comment]
There's a lot of getting the variable by name all the time. I think this was operating on the assumption that the variable number might change once the CDF is open, but that might not be a reasonable assumption. In practice I've found that the variable number is basically the order in which the variables are created in the file. If I can drop that assumption and do everything by var number, a lot of the code across all of pycdf would speed up quite a bit (and a lot of code would be dropped.) Going to look into this.

**Comment by jtniehof** [comment]
Development is happening in [jtniehof/spacepy/pycdf_attrs_speedup](/jtniehof/spacepy/tree/pycdf_attrs_speedup)

**Comment by jtniehof** [comment]
Ew. If a variable is deleted all the variables with numbers higher than it are immediately renumbered. Things are a lot faster without having to look up number by name all the time, but it's probably necessary.

We *could* just say "don't assign zVars to Python variables and then go mucking in the CDF" but I'm not sure about that either. Interested parties should look in the dev branch...it has a test that fails with the var-by-number assumption.

**Comment by jtniehof** [comment]
I kicked the idea around with one of my students who's been doing a lot of the CDF work on the user end and we're leaning towards proceeding with track-by-number with big caveats to not assign CDF variables to Python variables and then delete CDF variables while expecting the Python variables to work. This is what will break:

```
var2 = cdf['Var2']
del cdf['Var1']
print(var2[...]) #This is probably not actually Var2 anymore
```

An alternative is to continue to track-by-name (as it has been in the past) but maintain a cache of name-number mappings that gets invalidated when a variable is deleted. That's a fair bit of code to maintain the cache but it avoids diving into the CDF library. The pathology here is:

```
var2 = cdf['Var2']
del cdf['Var2']
print(var2[...]) #this will fail
cdf['Var2'] = newdata
print(var2[...]) #this will print the new data not the old
```

**Comment by balarsen** [comment]
I prefer reasonable idiot checking. As soon as you implement something that
has to have users be smart someone will do the bad thing.

Only an issue if in write mode. Something that makes sense in read mode
that would be faster?

**Comment by jtniehof** [comment]
Unfortunately the application where I'm really up the creek on this one is writing, so I can't just switch based on readonly status (the thought did occur to me, but that also involves a fair bit of either-or code).

I agree on idiot protection where possible. The current approach certainly puts in a lot more safety checks than the main CDF library, but it is a higher-level interface as well so that makes some sense.

The only way to be 100% sure is to have a version number in the CDF object and also in the variables, and whenever a destructive change occurs, bump the CDF version and make all the existing variables break.

If tracking variables by name is acceptable (given the pathology above), I'll go with name-tracking with cacheing of numbers and see how bad that is.

**Comment by jtniehof** [comment]
The cacheing is in place and reasonably fast. It's still reference-by-name and there are two unit tests marked as expected failure highlighting this pathology. Going to try to figure something smart but if I can't, will just have to document.

**Comment by jtniehof** [comment]
Opened #93 to capture the weirdness of reference-by-name. The PR that ultimately closes this issue will have some documentation on that front but won't "fix" it.

**Comment by jtniehof** [comment]
Closed by #94.
